### PR TITLE
[Feature] Add HttpOnly flag for OSF, API and ADMIN cookie [#OSF-6339, #SEC-16]

### PR DESCRIPTION
**Pull request by cslzchen**
## Purpose

OSF does not set `HttpOnly` flag. Using the flag when generating a cookie helps mitigate the risk of client side script accessing the protected cookie.

## Changes

```
if response is not None:
    response.set_cookie(
        settings.COOKIE_NAME,
        value=cookie_value,
        domain=settings.OSF_COOKIE_DOMAIN,
        secure=settings.SESSION_COOKIE_SECURE,
        httponly=settings.SESSION_COOKIE_HTTPONLY)
    return response
```

[
osf-cookie-httponly
](url)

Note: The above screenshot is from local environment. `Secure` flag is disabled locally.

## Side effects

Not compatible with IE9 or earlier. We do not support IE9.
Scripts can no longer access OSF cookie.

## Staging, Staging2, Production, Test Deployment

```
SESSION_COOKIE_HTTPONLY = True
```

## Local Development

Do not change `SESSION_COOKIE_HTTPONLY` to `False`. OSF cookie should never be accessed with scripts directly. Use Cross-Domain XMLHttpRequest with correct security settings instead.

## Tickets

https://openscience.atlassian.net/browse/OSF-6339
https://openscience.atlassian.net/browse/SEC-16

## PRs

Blocked by https://github.com/CenterForOpenScience/osf.io/pull/5506

### CODE REVIEW

File: framework/auth/cas.py [hunk 1]
```diff
@@ -208,15 +213,23 @@ def get_logout_url(*args, **kwargs):
...
+
 def make_response_from_ticket(ticket, service_url):
-    """Given a CAS ticket and service URL, attempt to the user and return a proper
-    redirect response.
     """
+    Given a CAS ticket and service URL, attempt to the user and return a proper redirect response.
+
+    :param ticket:
+    :param service_url:
+    :param user_agent:
```
Comment (mfraezz) [inline framework/auth/cas.py]: It doesn't look like `user_agent` is actually a param for this function.

File: framework/sessions/__init__.py [hunk 2]
```diff
@@ -95,7 +95,7 @@ def set_session(session):
     sessions[request._get_current_object()] = session
 
 
-def create_session(response, data=None):
+def create_session(response, data=None, user_agent=None):
```
Comment (mfraezz) [inline framework/sessions/__init__.py]: It doesn't look like `user_agent` is used in this function, or ever passed in when it's called

File: framework/auth/cas.py (hunk 1 above)
Comment (cslzchen) [inline framework/auth/cas.py]: You are right. Working on a clean-up right now. I used to have it for IE9 compatibility.

File: framework/auth/cas.py [hunk 3]
```diff
@@ -208,15 +213,23 @@ def get_logout_url(*args, **kwargs):
     """
     return get_client().get_logout_url(*args, **kwargs)
 
+
 def get_profile_url():
     """Convenience function for getting a profile URL for a user.
     """
     return get_client().get_profile_url()
 
+
```
Comment (mfraezz) [inline framework/auth/cas.py]: Nitpicky, but we tell `flake` to ignore the PEP8 requirement of having two blank lines between methods. Not a blocker

Comment (cslzchen) [inline framework/auth/cas.py]: Agreed. However, In code that have both 1 and 2 blank lines, do we prefer to make the style consistent or just leave them the way it is.

File: framework/sessions/__init__.py [hunk 4]
```diff
@@ -128,60 +122,53 @@ def before_request():
...
         user = get_user(
             email=request.authorization.username,
             password=request.authorization.password
         )
-        # Create empty session
+        # Create an empty session
         # TODO: Shoudn't need to create a session for Basic Auth
-        session = Session()
-        set_session(session)
+        user_session = Session()
```
Comment (cslzchen) [inline framework/sessions/__init__.py]: Note for CR: change shadowing variable name to better alternative.

File: framework/auth/cas.py (hunk 3 above)
Comment (mfraezz) [inline framework/auth/cas.py]: It's really up to personal preference, but according to [PEP8](https://www.python.org/dev/peps/pep-0008/#blank-lines):

> Surround top-level function and class definitions with two blank lines.
>
>Method definitions inside a class are surrounded by a single blank line.
>
>Extra blank lines may be used (sparingly) to separate groups of related functions. Blank lines may be omitted between a bunch of related one-liners (e.g. a set of dummy implementations).
>
>Use blank lines in functions, sparingly, to indicate logical sections.

**Comment by cslzchen** [comment]
Code clean up done:
- fix imports order: python, third-party, our own
- remove unused params `user-agent`
- unify comments style and blank lines
- rename three shadowing variables `session` to respective names

File: framework/auth/cas.py (hunk 3 above)
Comment (cslzchen) [inline framework/auth/cas.py]: Thanks. I am for the style as well so I try to improve them on files (only related to my tickets) Changing too much may cause headaches of merge conflicts for everyone so I took this "little by little" approach. :)

**Comment by mfraezz** [comment]
This seems to have a lot of extraneous formatting changes; the only thing that looks like it should affect behavior is the toggling of a `SESSION_COOKIE_HTTPONLY` flag (as expected). While the whitespace changes aren't all to my personal taste for readability, I applaud the docstring improvements. 👏 

Pass complete, pushing this ahead 👓

**Comment by cslzchen** [comment]
Pull from develop to solve merge conflicts. Travis error is unrelated to this PR.

**Pull request merged**